### PR TITLE
Implementing initial set of MRTK 2.0 touch-ready buttons

### LINKED ISSUE #9522: [XR] Implement MRTK 2.0-style touch-based buttons
Create support for buttons that can be directly interacted with, and upgrade the current button set from MRTK 1 -> MRTK 2. New buttons will be added alongside the MRTK 1-style buttons, rather than replace them, and use the same set of pointer events, sending pointer down, update, up in a similar fashion to traditional interactions

This will also accomplish most of the work for #8117

### PR DESCRIPTION
**Pull request by rickfromwork**
This change adds a new touch button class that extends Button3D by introducing fast touch-checking logic. Touch-enabled buttons can then extend this class with very little variation from a touch-disabled button.

Aside from the core touch button class, this PR introduces MRTK 2.0's touch-ready HolographicButton, and a touch-ready MeshButton3D, with similar characteristics to the existing buttons under a similar name (the new button names are prefixed with 'touch'). The visuals for the new HolographicButton have been updated to match those seen in MRTK 2.0, as well.

Right now, to have an input interact with the buttons via touch, one must introduce a new mesh, created on the utility layer, with the tag "touchEnabled." Once that is done, the touch buttons will respond to the mesh as it interacts with the button. This mesh is added to hand-tracked input in this PR.

Events fired from collisions match existing events - pointer enter/exit events are fired when the input gets within hover distance from the button, and press/release events are fired when the input gets close enough to touch/press the button. The positional data returned is equal to the closest point on the button from the input source at the time of the event.

This change also modifies Control3D to account for multiple pointers entering/exiting a single control, and having two pointers with the same ID on a control. This is to fix issues caused by the fact that the pointer ID for some inputs defaults to 0, and that far and near interactions can be triggered at the same time from the same input source/ID, depending on the angle of the far interaction's targeting beam.

I would like to also account for the size of the input mesh when computing its distance to the button, but short of using the bounding sphere and overestimating the size of the input collidable, I'm not sure how to do so.

This resolves #9522

### CODE REVIEW

**Comment by sebavan** [comment]
Looks pretty solid to me :-) Thanks a lot for the contrib. I will let @RaananW and @bghgary provide more insights.

You currenlty have a couple lint issues:

image

As well as code doc issues:

image

And finally missing a line in dist/preview release/what s new.md to credit yourself for the wonderful work :-)

**Review by RaananW — COMMENTED** [review]
Hey! This looks great :-)

Added a few code comments. In general I can summarize them like this:

1) Use arrow-functions instead of regular function() to keep context
2) Make sure all objects you create are being disposed
3) If you create objects on each frame, try reusing existing elements as much as possible to avoid constant GC updates
4) use `let` or `const` instead of var (this is nit-picking :-) )

I think linting and documentation failed as well, can you check in the CI log what the problem exactly and fix it?

Thanks!!

**Comment by RaananW** [comment]
oh, and...

WOOOHOOOOOOO!!!!!!!!!!

File: gui/src/3D/controls/touchButton3D.ts
Comment (rickfromwork) [reply gui/src/3D/controls/touchButton3D.ts:37]: Oh! I originally had it in its own private function, but wasn't able to get a reference to `this` when I did that (`this` kept referencing the object calling the beforeRender callbacks).

File: gui/src/3D/controls/touchButton3D.ts
Comment (rickfromwork) [reply gui/src/3D/controls/touchButton3D.ts:29]: Oh duh, how'd I forget about Nullable

File: gui/src/3D/controls/touchButton3D.ts
Comment (rickfromwork) [reply gui/src/3D/controls/touchButton3D.ts:97]: I just wanted to make it a singleton

File: src/XR/features/WebXRControllerPointerSelection.ts
Comment (rickfromwork) [reply src/XR/features/WebXRControllerPointerSelection.ts:514]: Oh, I forgot to remove that comment! I figured out what the problem was - when I was creating a new mesh for collisions on the buttons, I wasn't setting up the metadata for it, which the pointer selection code was expecting.

File: gui/src/3D/gui3DManager.ts
Comment (chrisfromwork) [inline gui/src/3D/gui3DManager.ts:209]: nit: are we expecting this collection to change as we iterate over it?

File: gui/src/3D/controls/touchButton3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/touchButton3D.ts:51]: >let [](start = 8, length = 3)

nit: it seems like this value doesn't change after construction. may want to use const instead

File: gui/src/3D/controls/touchButton3D.ts
Comment (chrisfromwork) [reply gui/src/3D/controls/touchButton3D.ts:97]: Why do we want this to be a singleton?

I'm not sure how object construction works in typescript, but it looks like in its current form the TouchButton3DManager is reconstructed every time a button is created. Do we know for sure that all buttons end up using the same singleton with this setup?

---
In reply to: [545277195](https://github.com/BabylonJS/Babylon.js/pull/9620#discussion_r545277195) [](ancestors = 545277195)

File: gui/src/3D/controls/touchButton3D.ts
Comment (chrisfromwork) [inline gui/src/3D/controls/touchButton3D.ts:163]: >(thi [](start = 31, length = 4)

nit: Is there any reason that there are additional parenthesis around properties in this if statement?

File: src/XR/features/WebXRHandTracking.ts
Comment (chrisfromwork) [inline src/XR/features/WebXRHandTracking.ts:96]: >sceneForNearInteraction [](start = 8, length = 23)

does the scene need to now be specified when initializing this feature? Is it possible to switch scenes while xr/hand tracking is still running?

File: src/XR/features/WebXRHandTracking.ts
Comment (chrisfromwork) [inline src/XR/features/WebXRHandTracking.ts:274]: >4 [](start = 73, length = 1)

nit: can this 4 be some other constant?

File: gui/src/3D/controls/touchButton3D.ts
Comment (RaananW) [reply gui/src/3D/controls/touchButton3D.ts:37]: setting the variable to equal an arrow function will solve this issue. if you want to solve it with a  private function you can use bind, but that would make little sense in this scenario

File: src/XR/features/WebXRHandTracking.ts
Comment (rickfromwork) [reply src/XR/features/WebXRHandTracking.ts:96]: The scene has to be passed in, because nothing outside of the GUI3DManager has knowledge about the utilityLayer scene that UI is created on.

Right now, as far as I'm aware, the UI scene is overlaid on top of the main scene. So if the main scene changes, the utility layer scene (the one being passed in) will stay the same.

File: gui/src/3D/controls/touchButton3D.ts
Comment (rickfromwork) [reply gui/src/3D/controls/touchButton3D.ts:97]: To be more precise, I wanted something static so that there was a single manager that had knowledge of every button. This is so I can do all of the collision checking within a single function call, allowing me to only call scene.registerBeforeRender once.

I found that having each button register independently caused a masive performance drop in my test scene, since I had 60+ buttons, and found it was directly related to the number of calls to registerBeforeRender.

Though now I'm wondering if I'd still see that drop if I used registerAfterRender...
Maybe someone more knowledgeable on performance costs can tell me what the best way to do this is, given that I'm okay with a frame delay.

Comment (bghgary) [reply gui/src/3D/controls/touchButton3D.ts:97]: What happens when there are multiple scenes?

File: gui/src/3D/controls/touchHolographicButton.ts
Comment (satyapoojasama) [reply gui/src/3D/controls/touchHolographicButton.ts:291]: Yes and No to that, as I see it this is created only once if sharedmaterials is true, which is set to true by default. If not, it is created more than once, but I see it disposed in the dispose function. Do I still need to be worried?

File: gui/src/3D/controls/touchHolographicButton.ts
Comment (satyapoojasama) [reply gui/src/3D/controls/touchHolographicButton.ts:87]: I don't think so, also this part of the code is copied from the HolographicButton.ts and I have not made any changes to it.

File: gui/src/3D/gui3DManager.ts
Comment (satyapoojasama) [reply gui/src/3D/gui3DManager.ts:211]: I am not sure, you think we might need a lock of some kind?

Comment (chrisfromwork) [reply gui/src/3D/gui3DManager.ts:211]: no, i just don't think we need to check whether the materialName is in the collection when we are iterating over the collection :)

---
In reply to: [545508015](https://github.com/BabylonJS/Babylon.js/pull/9620#discussion_r545508015) [](ancestors = 545508015)

File: src/XR/features/WebXRHandTracking.ts
Comment (chrisfromwork) [reply src/XR/features/WebXRHandTracking.ts:96]: if this object exists across multiple scenes, would we want logic for setting/removing the scene to check for collisions? Would someone ever want to specify multiple scenes containing buttons?

---
In reply to: [545398931](https://github.com/BabylonJS/Babylon.js/pull/9620#discussion_r545398931) [](ancestors = 545398931)

**Comment by sebavan** [comment]
@bghgary Let me know when to do a final check and merge on this one :-)

File: gui/src/3D/controls/touchButton3D.ts
Comment (rickfromwork) [reply gui/src/3D/controls/touchButton3D.ts:95]: I'm not sure. I don't know much about multi-scene setups in babylon.js, but so long as the original scene exists, I imagine this will point to the GUI scene for it. Though now I'm wondering how hand meshes/visuals work across multiple scenes as well, since I believe them to also be attached to the original scene on creation.

**Comment by sebavan** [comment]
Or @RaananW if you have some time to double check the fixes :-)

Else we ll probably merge it in Jan if there is no rush on your side @rickfromwork  ?

**Comment by bghgary** [comment]
@rickfromwork I've marked this PR as draft until we are ready to merge.

File: gui/src/3D/controls/touchHolographicButton.ts
Comment (RaananW) [reply gui/src/3D/controls/touchHolographicButton.ts:87]: we need to be sure that the materials are being disposed correctly when the component is disposed.  if it is not done there it is an error in the holographic button as well :-)

File: gui/src/3D/controls/touchHolographicButton.ts
Comment (RaananW) [reply gui/src/3D/controls/touchHolographicButton.ts:291]: you can check if it already exists before, and dispose it if it does. otherwise we will have those reserved resources to worry about.

**Comment by RaananW** [comment]
i will try finding time in the next fw days, but i think january will be the safest :-)

File: gui/src/3D/controls/touchButton3D.ts
Comment (RaananW) [reply gui/src/3D/controls/touchButton3D.ts:95]: This is a good question which I need to investigate. I believe you will need to init the hand meshes again for them to work in a new scene. 

Back to the original question, I still don't understand the need of a static private member. removing the static and using _buttonManager should have the same effect, no?

**Comment by rickfromwork** [comment]
> @rickfromwork I've marked this PR as draft until we are ready to merge.

That's fine with me. I wasn't actually sure on when to take it out of draft, and worried I took it out too early.

**Comment by RaananW** [comment]
The conflict is a simple documentation issue. I am fine with this getting out of draft. There are a few comments that need to be addressed, but otherwise LGTM

File: gui/src/3D/controls/touchButton3D.ts
Comment (rickfromwork) [reply gui/src/3D/controls/touchButton3D.ts:95]: @RaananW If I remove the static, then wouldn't each button have their own manager? That would mean multiple calls to scene.registerBeforeRender, which I'm trying to avoid by using a static manager.
I have found that excessive calls to this cause a large performance drop, at least on HoloLens. I noticed this when I copied the button panel playground scene, which generates 60 buttons.

File: gui/src/3D/gui3DManager.ts
Comment (rickfromwork) [reply gui/src/3D/gui3DManager.ts:246]: That's a good point... 
@deltakosh As the one who added this check back in 2018 to this._sharedMaterials, do you happen to remember what the check was for?

File: src/XR/features/WebXRHandTracking.ts
Comment (rickfromwork) [reply src/XR/features/WebXRHandTracking.ts:96]: I think that WebXRHandTracking doesn't work properly when switching scenes in general, since the hand mesh and joint meshes are all created on the current scene. It's something Raanan is going to look into, but the near interaction scene logic can be updated when the main hand scene logic is updated.

**Comment by RaananW** [comment]
LGTM! want to take it out of draft?

**Review by RaananW — APPROVED** [review]

**Comment by rickfromwork** [comment]
@sebavan @RaananW Should I merge this in, or did one of you want to do it/have any other questions before it can be merged?

**Comment by RaananW** [comment]
Merged! thanks a lot for that. can't wait to try the first playground with that :-)

**Pull request merged**
